### PR TITLE
Close connection when http acl rejected

### PR DESCRIPTION
**Pull request by linjunjj**

### CODE REVIEW

File: etc/emqx.conf [hunk 1]
```diff
@@ -423,6 +423,13 @@ acl_cache_max_size = 32
 ## Default: 1 minute
 acl_cache_ttl = 1m
 
+## Whether to enable disconnection when http acl rejected
+##
+## If enabled, ACL will close all connection
+##
```
[the remark this answers is not in the document]
Comment (linjunjj) [reply etc/emqx.conf]: sorry. I will correct..

File: src/emqx_protocol.erl [hunk 2]
```diff
@@ -873,6 +875,39 @@ parse_topic_filters(?SUBSCRIBE, RawTopicFilters) ->
...
     lists:map(fun emqx_topic:parse/1, RawTopicFilters).
 
+%% Issue#1641: not close connection when http acl rejected
+%%-----------------------------------------------------------------------------
+%% close connection when acl rejected and the config is available
+%%-----------------------------------------------------------------------------
+
+control_acl_pub_connection(PState) ->
+    case is_enable() of
+        true -> shutdown("close connection", PState);
```
Comment (terry-xiaoyu) [inline src/emqx_protocol.erl]: I don't think the this will disconnect your mqtt client. Have a test and make sure the connection is disconnected.

Comment (linjunjj) [reply src/emqx_protocol.erl]: yes! I realize it's mistake

Comment (linjunjj) [reply src/emqx_protocol.erl]: yes!  your are right

File: etc/emqx.conf (hunk 1 above)
[the remark this answers is not in the document]
Comment (Gilbert-Wong) [reply etc/emqx.conf]: If what enabled?

[the remark this answers is not in the document]
Comment (Gilbert-Wong) [reply etc/emqx.conf]: I suggest you to refer the description of other configuration items.

File: priv/emqx.schema [hunk 3]
```diff
@@ -543,7 +543,7 @@ end}.
 ]}.
 
 %% @doc Enable ACL close connection when acl rejected
-{mapping, "enable_acl_connection", "emqx.enable_acl_connection", [
+{mapping, "acl_disconnection_client", "emqx.enable_acl_connection", [
```
Comment (Gilbert-Wong) [inline priv/emqx.schema]: Maybe `acl_reject_connection` is better?

File: src/emqx_protocol.erl [hunk 4]
```diff
@@ -876,37 +876,24 @@ parse_topic_filters(?SUBSCRIBE, RawTopicFilters) ->
...
-    case is_enable() of
-        true -> shutdown("close connection", PState);
-        false -> {error, ?RC_NOT_AUTHORIZED}
-    end.
+try_close_connect(PState) ->
+     case acl_deny_disconnect() of
+         true -> PState1 = PState#pstate{connected = false},
+             {ok, PState1};
+         false -> {error, ?RC_NOT_AUTHORIZED}
+        end.
```
Comment (Gilbert-Wong) [inline src/emqx_protocol.erl]: The typo of `case` and `end`.

File: src/emqx_protocol.erl [hunk 5]
```diff
@@ -876,37 +876,24 @@ parse_topic_filters(?SUBSCRIBE, RawTopicFilters) ->
...
-                  username = Username,
-                  peername = Peername } .
 
 %%-----------------------------------------------------------------------------
 %%  get config
 %%-----------------------------------------------------------------------------
 
-is_enable() ->
+acl_deny_disconnect() ->
     application:get_env(emqx,enable_acl_connection,true).
```
Comment (Gilbert-Wong) [inline src/emqx_protocol.erl]: emqx_config:get_env().

File: priv/emqx.schema (hunk 3 above)
Comment (Gilbert-Wong) [reply priv/emqx.schema]: `acl_kick_client`

File: test/emqx_protocol_SUITE.erl [hunk 6]
```diff
@@ -17,72 +17,67 @@
 -module(emqx_protocol_SUITE).
 
 -compile(export_all).
+
```
Comment (Gilbert-Wong) [inline test/emqx_protocol_SUITE.erl]: Why do you have to add line ?

File: test/emqx_protocol_SUITE.erl [hunk 7]
```diff
@@ -17,72 +17,67 @@
...
 -include_lib("eunit/include/eunit.hrl").
 
 -include_lib("common_test/include/ct.hrl").
 
 -include("emqx_mqtt.hrl").
 
--define(TOPICS, [<<"TopicA">>, <<"TopicA/B">>, <<"Topic/C">>, <<"TopicA/C">>,
-                 <<"/TopicA">>]).
+-define(WILD_TOPICS,
```
Comment (Gilbert-Wong) [inline test/emqx_protocol_SUITE.erl]: These topic filters has been tested in other test suites, please don't add duplicated test suites.

File: src/emqx_protocol.erl [hunk 8]
```diff
@@ -341,19 +343,31 @@ process_packet(Packet = ?PUBLISH_PACKET(?QOS_0, Topic, _PacketId, _Payload), PSt
...
     case check_publish(Packet, PState) of
         {ok, PState1} ->
             do_publish(Packet, PState1);
+         {error, ?RC_NOT_AUTHORIZED} ->
+                    ?LOG(warning, "Cannot publish qos1 message to ~s for ~s",
+                        [Topic, emqx_reason_codes:text(?RC_NOT_AUTHORIZED)]),
+                    case deliver({puback, PacketId, ?RC_NOT_AUTHORIZED}, PState) of
+                      {ok, PState2} -> try_close_connect(PState2) ;
+                      {error, Reason} -> {error, Reason}
+                    end;
```
Comment (terry-xiaoyu) [inline src/emqx_protocol.erl]: Align this code block

File: etc/emqx.conf [hunk 9]
```diff
@@ -423,6 +423,13 @@ acl_cache_max_size = 32
 ## Default: 1 minute
 acl_cache_ttl = 1m
 
+## Disconnect when ACL denied
+##
+## If enabled, disconnect the client when  acl denied
+##
+## Value: on | off
+acl_kick_connection = on
```
[the remark this answers is not in the document]
Comment (terry-xiaoyu) [reply etc/emqx.conf]: Like this: `#acl_kick_connection = off`

File: priv/emqx.schema [hunk 10]
```diff
@@ -542,6 +542,13 @@ end}.
   {validators, ["range:gt_0"]}
 ]}.
 
+%% @doc Disconnect when ACL denied
+{mapping, "acl_kick_connection", "emqx.acl_kick_connection", [
+  {default, on},
```
Comment (terry-xiaoyu) [inline priv/emqx.schema]: {default, off},

File: src/emqx_protocol.erl [hunk 11]
```diff
@@ -874,6 +902,19 @@ parse_topic_filters(?SUBSCRIBE, RawTopicFilters) ->
...
+         true -> {error, kick_client};
+         false -> {ok, PState}
+    end.
+
+%%-----------------------------------------------------------------------------
+%%  get config
+%%-----------------------------------------------------------------------------
+
+acl_deny_disconnect() ->
+     emqx_config:get_env(acl_kick_connection,true).
```
Comment (terry-xiaoyu) [inline src/emqx_protocol.erl]: acl_kick_connection defaults to `false`

File: test/emqx_protocol_SUITE.erl [hunk 12]
```diff
@@ -420,5 +667,7 @@ raw_send_serialize(Packet, Opts) ->
     emqx_frame:serialize(Packet, Opts).
 
 raw_recv_parse(P, ProtoVersion) ->
-    emqx_frame:parse(P, {none, #{max_packet_size => ?MAX_PACKET_SIZE,
-                                 version         => ProtoVersion}}).
+    emqx_frame:parse(P,
+		     {none,
+		      #{max_packet_size => ?MAX_PACKET_SIZE,
+			version => ProtoVersion}}).
```
Comment (terry-xiaoyu) [inline test/emqx_protocol_SUITE.erl]: Our coding style (maybe different to the common Erlang coding guideline) uses 4 spaces indentation.

Comment (terry-xiaoyu) [reply test/emqx_protocol_SUITE.erl]: Revert all your changes regarding to the code format.

File: priv/emqx.schema [hunk 13]
```diff
@@ -542,6 +542,13 @@ end}.
   {validators, ["range:gt_0"]}
 ]}.
 
+%% @doc Disconnect when ACL denied
+{mapping, "acl_kick_connection", "emqx.acl_kick_connection", [
+  {default, off},
+  {datatype, flag}
+]
+}.
```
Comment (Gilbert-Wong) [inline priv/emqx.schema]: ]}.

File: src/emqx_protocol.erl [hunk 14]
```diff
@@ -427,8 +456,19 @@ process_packet(?SUBSCRIBE_PACKET(PacketId, Properties, RawTopicFilters),
...
+                    case deliver({suback, PacketId, ReasonCodes}, PState) of
+                        {ok, PState2} ->
+                            case acl_should_disconnect() of
+                                true -> {error, kick_client};
+                                false -> {ok, PState2}
+                            end;
+                        {error, Reason} ->  {error, Reason}
+                    end;
+                false -> deliver({suback, PacketId, ReasonCodes}, PState)
+                end
```
Comment (Gilbert-Wong) [inline src/emqx_protocol.erl]: Typo problem.

File: test/emqx_protocol_SUITE.erl [hunk 15]
```diff
@@ -407,8 +407,67 @@ subscribe_v5(_) ->
...
+            erlang:error("Client is not disconnected")
+    end,
+    %% 4. On sub ACL allowed
+    {ok, T4} = emqx_client:start_link([{host, "localhost"},
+        {client_id, <<"client4">>},
+        {username, <<"admin4">>},
+        {password, <<"pass4">>}]),
+    {ok, _} = emqx_client:connect(T4),
+    MSG = emqx_client:subscribe(T4, <<"a/b/c">>),
+    ct:pal("Connection closed: ~p~n", [MSG]).
```
[the remark this answers is not in the document]
Comment (Gilbert-Wong) [reply test/emqx_protocol_SUITE.erl]: Please follow advice of @terry-xiaoyu

**Pull request closed without merging**